lang/rust/avro/src/codec.rs

```diff
@@ -49,7 +49,13 @@ pub enum Codec {
     /// CRC32 checksum of the uncompressed data in the block.
     Snappy,
     #[cfg(feature = "zstandard")]
+    /// The `Zstandard` codec uses Facebook's [Zstandard](https://facebook.github.io/zstd/) with the
+    /// default compression level.
     Zstandard,
+    #[cfg(feature = "zstandard")]
+    /// This codec is the same as `Zstandard` but allows specifying the compression level.
+    #[strum(disabled)]
+    ZstandardWithLevel(ZstandardLevel),
```

Comment:
Since we are in `0.x` state, I don't think we need to avoid breaking changes. Directly changing `Zstandard` looks cleaner.

Reply:
I agree!
I'd suggest to go even further:
1) introduce ZstandardOptions (or ZstandardConfig) struct that provides all possible settings - the level and a dictionary. With a `Default` impl that uses `0` and `&[]`
2) the same for all other codecs with their respective settings

WDYT ?

Reply:
With https://github.com/apache/avro/pull/3095/commits/71021f21185634ffa05861b931f5416c481ee90a I've pushed codec settings for Bzip2. Please pull before working on the settings for Zstd!

Reply:
Excellent idea! I'll go ahead with it for zstd.

Reply:
I also pushed similar settings for Xz!

Snappy does not provide APIs for setting compression level (or anything other), so there is no need of changes!

Deflate provides its own `EncodeOptions` but we cannot use it in our `Codec` enum because it does not implement several traits (Clone, Copy, Eq and PartialEq) ... Constructing EncodeOptions from its fields is not possible because it does not provide such APIs.

So, only Zstd is left! I'll leave it to you @MarcoLugo !

Reply:
I added a similar struct for zstd but for now I ended up leaving out the `dictionary` as it would be required for decoding as well and so it would need to go into the header metadata and I'm not sure how to do that at the moment. I'll need to read the avro spec in more detail, maybe this could be a future PR in order to avoid blocking this one?
I also made the new settings structs public so they're visible in the docs and can be used.

Reply:
On second thought, maybe the dictionary option can still be provided and left to the user for now without it necessarily being seamless (i.e. the dev is responsible for providing it at both encoding and decoding time) if eventually this can be made even more seamless by just being necessary at encoding time then that's a nice bonus.

Reply:
The dictionary is a Zstd thingy. It just needs to be the same for both encoding and decoding.
Just pushed https://github.com/apache/avro/pull/3095/commits/8f4a2f3c681d473a8dc62cd48aa5a93ff2e42e50 for it

Reply:
I am going to revert the support for dictionary. Dealing with the slice causes a lot of lifetime headaches.
I am about to commit an improvement to the (de)ser of the codec in the Avro binary metadata. We need to store the settings along with the codec name!

Reply:
I think I hit a problem!

By Avro spec when writing the binary data (.avro) the codec is stored in the header's metadata as a string, e.g. "zstandard". But there is nothing about codec's metadata!
So, if you do something like :
```
let schema = Schema::parse_str(
        r#"
        {
            "type": "record",
            "name": "Test",
            "fields": [
                {"name": "f1", "type": "int"},
                {"name": "f2", "type": "string"}
            ]
        }"#,
    )?;

    let mut writer = Writer::with_codec(&schema, Vec::new(), codec);
    let mut record = Record::new(writer.schema()).unwrap();
    record.put("f1", 27_i32);
    record.put("f2", "foo");
    writer.append(record)?;
    let input = writer.into_inner()?;
    let mut reader = Reader::new(&input[..])?;
    ...   
```
Then the Reader will not know anything about the codec's settings used by the writer! The Reader will use the codec with its default settings!
We can store the settings in the metadata's `user_metadata`, see reader.rs > Block struct. But this will take some more time!

Reply:
This is what I was fearing earlier regarding the dictionary though I am admittedly less familiar with the spec. I'd be in favor of having the zstd dictionary as a separate ticket/PR. This ticket initially it was about being able to set the level which I think is already a good improvement and we got the ability to set the level for more codecs than zstd too :)